1032 allow referred search terms

juzen2003 opened the pull request:
- Fixes #1032 
- Were any Django, import pipeline, table_schema, or dictionary files modified? Y
  - If YES:
    - Database used: import on an empty database 
    - All Django tests pass: Y
    - FLAKE8 run on modified code: Y
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y (menu.js)
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
- In search sidebar, select metadata menu:
    - Add Observed Ring Radius under Occultation Constraints
    - Add Exposure Duration under Cassini ISS
    - Add Ring Center Opening Angle to Sun under Occultation Constraints
    - Add Ring Center Opening Angle to Observer under Occultation Constraints

Known problems: None

juzen2003 committed:
Add referred_slug field in param_info table schema, and add two more
fields in obs_occultation to point to two ring radius referred slugs in
param_info.
- In param_info.json, add referred_slug field. (line 55-57)
- In obs_occultation.json, add two ring radius fields pointing to referred_slug
  field in param info. (line 345-361)
- In do_param_info.py, put referred_slug as a field in the new row and
  allow None in other fields for param info. (line 95-106)

juzen2003 committed:
Make the corresponding slug available under the category that has referred_slug
field in search sidebar and select metadata menu.
- In views.py, in _get_menu_labels, add the corresponding slug under the
  current category or sub-category when referred_slug field is available.
  (line 1524-1530, 1548-1553)
- In models.py, add referred_slug field in ParamInfo model. (line 28)

juzen2003 committed:
Modify body_qualified_label_results and body_qualified_label to make sure
"[Ring]" or "[<Surface Body>]" or "[Mission]" or "[Instrument]" is not
duplicated in the label for referred slug.
- In models.py, if pretty_name exists in the label, we don't append it
  again before returning. (line 77, 96-99)

juzen2003 committed:
Modify search sidebar menu item click event handler to highlight and
de-highlight the labels pointing to the same widget.
- In menu.js, update the selector passed into o_menu.markMenuItem so that
  it will highlight and de-highlight the labels that have the same
  data-slug. (pointing to the same widget) (line 33-34)

juzen2003 committed:
Add ring_radius1 & ring_radius2 (two field name of referred slug) in
ObsOccultation model in models.py (line 1770-1771)

juzen2003 committed:
Change referred slug (to ring radius) field names to referred_ring_radius1
and referred_ring_radius2 in occultation table.
- In models.py, change ObsOccultation referred slug model names to referred_ring_radius1
and referred_ring_radius2. (line 1770-1771)
- In obs_occultation.json, change referred_slug field name to referred_ring_radius1
and referred_ring_radius2. (line 345, 354)

juzen2003 committed:
Fix python test and change referred_slug field names to the same ones as
the field names of it's original tables. (ring_radius1 & ring_radius2)
- In models.py (line 1770-1771) and obs_occultation.json (line 345,354),
  change field names back to ring_radius1 & ring_radius2.
- In api_metadata_vg_iss_2_s_c4360845_default_html.html (line 64-65) and
  api_metadata_vg_iss_2_s_c4360845_default_json.json (line 56-57), fix
  the expected results for python tests.
- In do_param_info.py, removed commented out codes at line 100.

juzen2003 committed:
In COISS, add a referred_slug field to exposure duration.
- In obs_instrument_coiss.json, add a field for referred slug to exposure
  duration. (line 419-427)
- In models.py, add duration to the model of ObsInstrumentCoiss (line 1457).

juzen2003 committed:
Add referred_slug fields to solar/observer ring opening angle in occultation
table.
- In obs_occultation.json, add fields for referred slug to observer_ring_opening_angle
  and solar_ring_opening_angle. (line 363-398)
- In models.py, add solar_ring_opening_angle1/2 and observer_ring_opening_angle1/2
  to the model of ObsOccultation (line 1457).

juzen2003 committed:
Fix python test.
- In results/views.py, make sure proper param info is stored in all_info
  so that the label_results of param info of exposure duration is not
  None. (line 590-595)
- In api_metadata_vg_iss_2_s_c4360845_default_html.html (line 66-69) &
  test_api/responses/api_metadata_vg_iss_2_s_c4360845_default_json.json
  fix the expected results. (line 49-50, 60-61)
- In ui/views.py, update the comments. (line 1524, 1525)

juzen2003 committed:
In table schemas, modify the pi_disp_order for referred_slugs to be 1000+.

juzen2003 committed:
Add exposure duration in the table schemas of Galileo SSI, VGISS, Hubble
, and New Horizons LORRI.

juzen2003 committed:
Add duration to the models of ObsInstrumentGossi, ObsInstrumentNhlorri,
ObsMissionHubble, ObsInstrumentVgiss.

juzen2003 committed:
Fix python test for hubble and vgiss after adding exposure duration as
the referred field.

rfrenchseti commented on opus/application/apps/search/models.py:
```diff
@@ -1454,6 +1454,7 @@ class ObsInstrumentCoiss(models.Model):
     target_desc = models.CharField(max_length=75)
     combined_filter = models.CharField(max_length=30)
     camera = models.CharField(max_length=1)
+    duration = models.FloatField(blank=True, null=True)
```
There should be no changes to models.py (either here or in ObsOccultation below). We aren't actually adding new fields to the database.

rfrenchseti commented on opus/application/test_api/responses/api_metadata_vg_iss_2_s_c4360845_default_html.html:
```diff
@@ -61,6 +61,12 @@
 <dt>Wavelength Band</dt><dd>N/A</dd>
 <dt>Signal Source Name</dt><dd>N/A</dd>
 <dt>Receiver Host Name</dt><dd>N/A</dd>
+<dt>Observed Ring Radius (Min) (km)</dt><dd>N/A</dd>
```
These should have a [Ring] after them in the resulting metadata, not just on the search page.

rfrenchseti commented on opus/import/table_schemas/obs_instrument_coiss.json:
```diff
@@ -415,6 +415,15 @@
+        "field_type": "real8",
+        "pi_category_name": "obs_instrument_coiss",
+        "pi_referred_slug": "duration",
+        "pi_disp_order": 100,
```
pi_disp_order needs to be adjusted for all referred slugs to appear after all other disp_orders for this category (we want the referred slugs to be at the bottom of the list.

rfrenchseti reviewed (changes requested):
- On the Search tab, referred slugs show up like "Exposure Duration [Image]", which is correct. But in the Select Metadata dialog they just say "Exposure Duration". The qualifier should be included there as well.
- Referred fields are not showing up on the Detail tab at all (in their new location).

juzen2003 committed:
Add a link icon at the end of referred menu items.

juzen2003 committed:
Remove referred slug columns in the database.
- Remove field_name, field_type of referred_slug columns in table schemas
  json files.
- In paraminfo/models.py, modify body_qualified_label_results so that
  "[Image]" is appended at the end of menu label in select metadata menu.
  (line 93)
- In mysql.py, add logic to make sure we don't check field_name and field_type
  for referred_slug entries. (line 317,339)
- In search/models.py, remove all referred slug models.
- In do_import.py, make sure we don't check referred_slug entry when creating
  mult table. (line 212)

juzen2003 committed:
Make referred slug metadata show up in detail tab by modifying get_metadata.
- In views.py, in get_metadata, modify the logic to obtain result from
  _get_metadata_by_slugs for referred_slug entries. (line 626-634, 647-659)

juzen2003 committed:
Fixed test files of results_.*.json

juzen2003 committed:
Modify get_metadata to store all param info objects in a dictionary keyed
by table label. This is to distinguish the referred slug param info and
the original param info in different categories.
- In views.py (line 555, 608-609), store all param info objects in a
  dictionary data_all_info keyed by table label. This dictionary will be
  passed to django metadata template later.
- In detail_metadata.html (line 6-16), detail_metadata_internal.html (line
  6-31), add for loop to obtain proper all info based on current category.

juzen2003 committed:
Fixed test files of api_metadata_.*.json/html/csv & api_metadata2_.*.json/html/csv.

juzen2003 committed:
Modify get_fields_info to change the return data structure by adding another
level (category) for slugs.
- In views.py, include referred slug in the return. (line 709-719)
- In views.py, add category as another level to keyed the related slugs
  in the return. (line 727-729, 778)
- In views.py, sort the data before returning. (line 780-791)

juzen2003 committed:
Modify get_fields_info to include referred slug in .csv return, and also
add "referred" field in the return data.

juzen2003 committed:
Fixed apiguide.html test and modify api_api_guide to properly fetch fields
data from get_fields_info.
- In views.py (line 312-318), modify api_api_guide to properly fetch fields
  data from get_fields_info.

juzen2003 committed:
Modify API guide and modify get_fields_info to have proper category for
surface geometry in return data when collapse is True.

juzen2003 committed:
Restore the import from tools.db_utils in views.py.

juzen2003 committed:
Update styling for link icon for referred slug menu item.

rfrenchseti commented on opus/import/importdb/mysql.py:
```diff
@@ -334,6 +335,8 @@ def create_table(self, namespace, raw_table_name, schema,
                 mult_schema.append(entry)
             new_schema = []
             for column in schema:
+                if column.get('pi_referred_slug', None):
```
This should be False not None

rfrenchseti commented on opus/application/apps/help/api_guide.md:
```diff
@@ -1790,53 +1790,59 @@ Examples:
-    "rightasc2": {
-      "label": "Right Ascension (Max)",
-      "search_label": "Right Ascension",
+    "obs_general": {
```
Use full name like "General Constraints" instead of internal table names.

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -581,19 +581,20 @@ def api_get_fields(request, fmt, slug=None):
-          "old_slug": "timesec1",
-          "slug": "time1"
-        }
+        "obs_general"
```
Update with full category name

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -706,17 +707,35 @@ def get_fields_info(fmt, request, api_code, slug=None, collapse=False):
                 # Internal use only
                 continue
+            cat = f.category_name
+            if collapse and cat.find('obs_surface_geometry__') != -1:
```
Can use startswith()

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -762,15 +781,21 @@ def get_fields_info(fmt, request, api_code, slug=None, collapse=False):
             entry['slug'] = entry['field_id'] # Backwards compatibility
-            return_obj[collapsed_slug] = entry
+            entry['referred'] = True if f.referred_slug else False
+            return_obj.get(cat)[collapsed_slug] = entry
```
Can't this just be return_obs[cat] instead of .get?

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -762,15 +781,21 @@ def get_fields_info(fmt, request, api_code, slug=None, collapse=False):
+        for cat, cat_data in return_obj.items():
+            del cat_data['table_order']
+        # Sort slugs of each category by disp_order
+        return_obj[cat] = OrderedDict(sorted(return_obj[cat].items(),
```
I'm confused. Shouldn't this be indented so it's part of the loop? In which case you don't need return_obj[cat] since that's just cat_data. The way it is now you're just sorting the last category. When this is fixed you can combine it with the loop below like it was before.

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -784,14 +809,19 @@ def get_fields_info(fmt, request, api_code, slug=None, collapse=False):
+        for cat, cat_data in return_obj.items():
+            for k, v in cat_data.items():
+                row_data = [(v['field_id'], v['category'], v['type'],
+                         v['search_label'], v['label'],
```
Indent properly

rfrenchseti commented on opus/application/apps/paraminfo/models.py:
```diff
@@ -88,8 +90,12 @@ def body_qualified_label_results(self):
-        if pretty_name in ['General', 'PDS', 'Wavelength', 'Image',
-                           'Occultation', 'Surface']:
+        if (pretty_name in ['General', 'PDS', 'Wavelength', 'Image',
+                           'Occultation', 'Surface'] and not referred):
```
Indent properly.

rfrenchseti commented on opus/application/apps/results/views.py:
```diff
@@ -612,6 +626,17 @@ def get_metadata(request, opus_id, fmt, api_name, return_db_names, internal):
+                                param_info.body_qualified_label_results(True))
+                    # Assign referred_slug, this will be used to determine if
+                    # the param info is from referred_slug, and we will used
+                    # the slug to get the metadata result later.
```
Add period after "Assign referred_slug" and start new sentence. Also used -> use

rfrenchseti commented on opus/application/test_api/responses/api_metadata_vg_iss_2_s_c4360845_default_html.html:
```diff
@@ -97,8 +103,8 @@
 <dt>Observed Emission Angle (Max) (degrees)</dt><dd>88.886</dd>
-<dt>Observed Solar Hour Angle (Min) (degrees)</dt><dd>0</dd>
-<dt>Observed Solar Hour Angle (Max) (degrees)</dt><dd>360</dd>
+<dt>Observed Local Hour Angle (Min) (degrees)</dt><dd>0</dd>
```
Why did this change?

juzen2003 replied:
Under <target> surface geometry (in obs_suraface_geometry_target.json), the label_results for "solar_hour_angle1/2" fields are "Observed Local Hour Angle (Min/Max)"

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -208,7 +208,8 @@ def create_tables_for_import(volume_id, namespace):
         for table_column in table_schema:
-            if table_column.get('put_mults_here', False):
+            if (table_column.get('put_mults_here', False) or
+                table_column.get('pi_referred_slug', None)):
```
Should be False instead of None

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1286,7 +1287,8 @@ def import_observation_table(volume_id,
     for table_column in ordered_columns:
-        if table_column.get('put_mults_here', False):
+        if (table_column.get('put_mults_here', False) or
+            table_column.get('pi_referred_slug', None)):
```
Same here

juzen2003 committed:
Update fields.json return, change "referred" to "linked", and also change
the frist level key of return data structure from category_name to table
label name.
- In views.py (line 730-734, 780), modify get_fields_info, change the
frist level key of return data structure from category_name to table
label name, and also change "referred" to "linked"
- In api_guide.md, update the return data of fields.json. (line 1827,1863)

juzen2003 committed:
Add link icon for referred fields in detail tab & "+" menu in slideshow
and table view.

juzen2003 committed:
Add tooltips for referred field link icon.
- In models.py, create get_link_tooltips to return the tooltips
  for referred field link icon (line 64-69).
- In detail_metadata_internal.html, add_field.html, menu.html, add title
  attribute to link icon i tag.

juzen2003 committed:
Add info (blue text) text in the widgets of Cassini ISS filter and Hubble
Filter Name.

juzen2003 committed:
Update code styles & minor logic change based on github comments.

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -581,19 +581,20 @@ def api_get_fields(request, fmt, slug=None):
+              "available_units": null,
+              "old_slug": "timesec1",
+              "slug": "time1"
+            }
```
Needs to have "linked" field in comment

rfrenchseti commented on opus/application/apps/help/api_guide.md:
```diff
@@ -1790,53 +1790,59 @@ Examples:
 %CODE%
 {
   "data": {
-    "planet": {
```
Need to add the "linked" field to the table above. It should go after the current last field (default_units) and say:

| `linked` | `true` if this field is not native to this category but has been linked from its normal location |

Also the table for [field].[fmt]:

| `linked` | Always `false` because this API call returns information about the field's native category  |

rfrenchseti commented on opus/application/apps/help/api_guide.md:
```diff
@@ -1790,53 +1790,59 @@ Examples:
+        "default_units": null,
+        "available_units": null,
+        "category": "General Constraints",
+        "field_id": "planet"
```
Need to add "linked" to these results for both fields API calls

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -762,15 +782,20 @@ def get_fields_info(fmt, request, api_code, slug=None, collapse=False):
-        for key, val in return_obj.items():
-            del val['table_order']
-            del val['disp_order']
+                                    key=lambda x: x[1]['table_order']))
```
Fix indentation

rfrenchseti commented on opus/application/apps/metadata/views.py:
```diff
@@ -762,15 +782,20 @@ def get_fields_info(fmt, request, api_code, slug=None, collapse=False):
+            del cat_data['table_order']
+            # Sort slugs of each category by disp_order
+            cat_data = OrderedDict(sorted(cat_data.items(),
+                                    key=lambda x: x[1]['disp_order']))
```
Fix indentation

rfrenchseti commented on opus/application/apps/results/views.py:
```diff
@@ -579,18 +580,33 @@ def get_metadata(request, opus_id, fmt, api_name, return_db_names, internal):
+                    param_info.label_results = (
+                                param_info.body_qualified_label_results(True))
+                    param_info.referred_slug = referred_slug
+
```
Remove blank line

rfrenchseti commented on opus/application/apps/paraminfo/models.py:
```diff
@@ -60,6 +61,12 @@ def get_tooltip_results(self):
             definition = get_def_for_tooltip(self.dict_name, self.dict_context)
         return definition
 
+    def get_link_tooltips(self):
```
Shouldn't this just be "get_link_tooltip"? There's only one tooltip for this paraminfo.

juzen2003 committed:
Update api_guide.md & comments and fix code styles based on github comments.

juzen2003 committed:
Include linked field in .csv return. Update the corresponding API guide
and test.

juzen2003 committed:
Fix indentation at line 813 in views.py to pass the flake8 test.

juzen2003 committed:
Update comments for api_get_fields, and update api guide style at line 1778.

juzen2003 committed:
Modify get_fields_info to store the sorted results of each category back
to the return objects at the end (line 799) in views.py.
